bgpd/bgp_nht.c

```diff
@@ -53,16 +53,28 @@ static void unregister_zebra_rnh(struct bgp_nexthop_cache *bnc,
 static void evaluate_paths(struct bgp_nexthop_cache *bnc);
 static int make_prefix(int afi, struct bgp_path_info *pi, struct prefix *p);
 
-static int bgp_isvalid_nexthop(struct bgp_nexthop_cache *bnc)
+static int bgp_isvalid_nexthop(struct bgp_path_info *path,
+			       struct bgp_nexthop_cache *bnc)
 {
-	return (bgp_zebra_num_connects() == 0
-		|| (bnc && CHECK_FLAG(bnc->flags, BGP_NEXTHOP_VALID)));
-}
+	if (bgp_zebra_num_connects() == 0)
+		return 1;
 
-static int bgp_isvalid_labeled_nexthop(struct bgp_nexthop_cache *bnc)
-{
-	return (bgp_zebra_num_connects() == 0
-		|| (bnc && CHECK_FLAG(bnc->flags, BGP_NEXTHOP_LABELED_VALID)));
+	if (path) {
+		safi_t safi = bgp_node_table(path->net)->safi;
+
+		/*
+		 * VPN routes and associated IP routes need an LSP to the BGP
+		 * next hop.
+		 */
+		if (safi == SAFI_MPLS_VPN
+		    || (safi == SAFI_UNICAST
+			&& path->sub_type == BGP_ROUTE_IMPORTED && path->extra
+			&& path->extra->num_labels))
+			return CHECK_FLAG(bnc->flags,
+					  BGP_NEXTHOP_LABELED_VALID);
+	}
```

Comment:
my understanding is that if we have an imp-null label on a route going to gre interface, then it is ok for the new bgp nht algorithm to go with that imp-null label. 
there are two ways to have this:
1- either enable ldp
2- or do a static route with nexthop and label imp-null going through gre

today, with frr, only 1 is possible.
creating 2 is not possible since imp-null label is forbidden for configuration on staticd.

Renato , do you confirm that the algorithm should be ok with one of the 2 exposed solutions ?
Doing a specific for gre interface ( not having imp-null) may not work because gre interface may also be point to multipoint interfaces.